### PR TITLE
file: allow to create CephFS dataPools with predefined names

### PR DESCRIPTION
**Pull request by degorenko**
Add an ability to specify CephFS dataPools as a map with a predefined
pool name.

Related-Issue: rook#9295
Signed-off-by: Denis Egorenko <degorenko@mirantis.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:968]: I assume it is not supported for users to specify both of these?

```suggestion
	// (Deprecated in favor of dataPoolsMapped) The data pool settings.
	// Exactly one of DataPools or DataPoolsMapped (preferred) must be specified.
```

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:973]: I assume it is not supported for users to specify both of these?

```suggestion
	// The data pool settings with explicit data pool name.
	// Exactly one of DataPools or DataPoolsMapped (preferred) must be specified.
```

File: pkg/operator/ceph/file/filesystem.go
Comment (BlaineEXE) [inline pkg/operator/ceph/file/filesystem.go:57]: I assume it shouldn't be supported for users to specify both of these. If so, we should also include a check to make sure that the user has not specified both DataPools and DataPoolsMapped. Does that make sense to you @travisn ?

File: pkg/operator/ceph/file/filesystem.go
Comment (BlaineEXE) [inline pkg/operator/ceph/file/filesystem.go:184]:
```suggestion
	for poolName, poolSpec := range spec.DataPoolsMapped {
		err := cephclient.CreatePoolWithProfile(context, clusterInfo, clusterSpec, poolName, poolSpec, "")
		if err != nil {
			return errors.Wrapf(err, "failed to update datapool  %q", poolName)
		}
	}
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
Overall, I think this seems good. It would be good to figure out how we can suggest users migrate from `dataPools` to `dataPoolsMapped`. 

I also don't see changes to the documentation (`Documentation/ceph-filesystem-crd.md` file). On that note, it would be good to mention how users can migrate from `dataPools` to `dataPoolsMapped` in a new section in the doc, I think.

Also, I wonder if `namedDataPools` might be a better name than `dataPoolsMapped`.

**Comment by degorenko** [comment]
@BlaineEXE yeah, i skipped changes to documentation just in case if overall approach will not suit community vision. Will add docs. For users we can raise info/warning message that old way is deprecated and they can switch to a new one. Also mention it in docs.

> Also, I wonder if namedDataPools might be a better name than dataPoolsMapped.

That was my first thought too :) Will update.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (degorenko) [reply pkg/apis/ceph.rook.io/v1/types.go:968]: It's discussable. TBH, give a user an ability to specify both at time will not lead to any problems. At least, we may check that pools specified in map are not intersect with pools from array. Wdyt?

File: pkg/operator/ceph/file/filesystem.go
Comment (travisn) [reply pkg/operator/ceph/file/filesystem.go:57]: There are a number of important design points that come to mind with this feature. What about the following?
- For backward compatibility, we should still support the `dataPools`. Let's keep it as a fully supported feature rather than deprecating it. Perhaps we eventually deprecate it, but with some unit tests it seems like it should be very simple to support. 
- Let's only allow either `dataPools` or `namedDataPools`. No need to support both in the same filesystem CR. The operator would fail the reconcile if both are specified.
- The example manifests and docs could all be updated to use namedDataPools by default.
- Ordering is important for adding the pools to cephfs. As seen in #8452, to support EC, a replica data pool must be created first, then the EC pool. So instead of a map we need to keep an ordered list. For the types.go, this might look like the following where the PoolSpec is an inline property:
```go
type NamedPoolSpec struct {
	Name string `json:"name,omitempty"`
	PoolSpec
}
type FilesystemSpec struct {
	DataPools []PoolSpec `json:"dataPools"`
	NamedDataPools []NamedPoolSpec `json:"namedDataPools"`
...
}
```

- For the implementation, before the ceph pools are created, we would translate all the dataPools into namedDataPools with the default pool names. Then all the pools can be processed the same. No need for a separate implementation in the operator code for creating the ceph pools.
- If users desired to update their CRs to the named pools, they would simply need to rename `dataPools` to `namedDataPools` and set the `name` of each pool to the name that was being generated previously by rook.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/file/filesystem.go
Comment (degorenko) [reply pkg/operator/ceph/file/filesystem.go:57]: In that case we may follow way:

```
type NamedPoolSpec struct {
	Name string `json:"name,omitempty"`
        Default bool `json:"default"`
	PoolSpec
}
type FilesystemSpec struct {
	DataPools []PoolSpec `json:"dataPools"`
	NamedDataPools []NamedPoolSpec `json:"namedDataPools"`
...
}
```

In that case we dont need an ordered list, because ordering seems to me is a hardcode and workaround. Before filesystem is going to be created check that default pool is present and it is the only one. For backward compatibility, since right now 1 element of array is assuming to be default we dont have any problems with transition to a new structure.

@travisn  @BlaineEXE what do you think?

Comment (travisn) [reply pkg/operator/ceph/file/filesystem.go:57]: Agreed, the designation of the first/default pool is the only thing cephfs cares about. The ordering of the additional data pools is unimportant to ceph. 

I still like the idea of keeping the ordered list. There are several advantages:
- No need for a new property to designate the default
- Follows the same pattern we've already been using, we're just adding a `name` property and renaming it to `namedDataPools`. 
- Very easy to convert existing filesystem CRs to the new approach.

Comment (travisn) [reply pkg/operator/ceph/file/filesystem.go:57]: Actually, we should be able to get away with not renaming it to `namedDataPools`, we should just be able to add the `name` to the existing pools. Then if the name is empty, we generate the name.

```go
type NamedPoolSpec struct {
	Name string `json:"name,omitempty"`
	PoolSpec
}
type FilesystemSpec struct {
	DataPools []NamedPoolSpec `json:"dataPools"`
...
}
```

Comment (travisn) [reply pkg/operator/ceph/file/filesystem.go:57]: After the initial creation of the filesystem, is there a way to change the primary pool? For example, if you remove a named pool, it wouldn't affect the other named pools being removed. Or what is the issue with pool removal?

If we use the inline flag, I believe the compatibility would be fine, it would just be handled by the json deserialization inside the operator.
```go
type NamedPoolSpec struct {
	Name     string `json:"name,omitempty"`
	PoolSpec `json:",inline"`
}
```

But another thought I'm having is that we could just add the `name` property to the PoolSpec. This would allow the user to specify pool names on the cephfs pools to override the default, and with #9209 for configuring the built-in ceph pool. It just wouldn't make sense in some areas like the rgw pool names, which I don't believe are configurable, so maybe that approach is less preferred.

Comment (degorenko) [reply pkg/operator/ceph/file/filesystem.go:57]: Ok, so if we are assuming that first element of `dataPools` array is always coming as default one (on creation) and we dont manage pool removal for cephfs, let's go your way.

Another reason for specifying `default` property is adding validation to avoid issue https://github.com/rook/rook/pull/8452 - give to a user an explicit error message what is going wrong without double checking docs. But since user is allowed to drop from spec any pool without its actual removal, then it will not work correctly. In the same time, we can add such validation on creation. Do you agree? I will post a comment in related issue.

I'll post update as discussed above, with auto transition to a new `NamedPoolSpec` structure, all pools present in current `dataPools` will be named by default pool name + its index.

Comment (travisn) [reply pkg/operator/ceph/file/filesystem.go:57]: Looks like there is a request on #8452 to not prevent the creation in that case. Instead of blocking the case, logging a warning seems sufficient.

> I'll post update as discussed above, with auto transition to a new NamedPoolSpec structure, all pools present in current dataPools will be named by default pool name + its index.

Agreed on this approach, thanks.

**Comment by degorenko** [comment]
@travisn @BlaineEXE updated. May be the last question, which we need to answer: do we need to create cephfs data pools with cephfs name prefix? Or name totally comes from spec?

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:977]: Based on the discussion, I was thinking we would stick with only the `dataPools` and effectively be converting all of the FS data pools into named pools by changing this type to NamedPoolSpec. Since it's only internal json serialization by the operator, to the CRD it simply looks like the `name` is added, and no CRD type conversion needed.
```suggestion
	DataPools []NamedPoolSpec `json:"dataPools,omitempty"`
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (degorenko) [reply pkg/apis/ceph.rook.io/v1/types.go:977]: In that case `name` parameter should be an optional for `NamedPoolSpec` - and if it is missed in spec we will use current default naming `<cephfsName>-data<index>`. So, then my question is still actual: do we need to create cephfs data pools with cephfs name prefix for `NamedPoolSpec`, e.g. `<cephfsName>-<nameFromSpec>` ?

Comment (degorenko) [reply pkg/apis/ceph.rook.io/v1/types.go:977]: Also, such behaviour forces users to add pools with names add the end of current list **only**, otherwise it may brake current naming.

Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go]: Correct, the `name` would be optional, and if blank rook would set the generated pool name as today. 

If the set the `name` in a filesystem CR that already exists, and it is different than the existing generated pool name, it would certainly create a problem so we should add a check for that case. Otherwise, setting the pool name would cause a new pool to be created and added to the filesystem, and the default-named pool(s) would remain. 

> Also, such behaviour forces users to add pools with names add the end of current list only, otherwise it may brake current naming.

If they add the `name`, maybe we should enforce that the `name` be set on all the pools in the list. This would allow them to be reordered, even for the existing default-named pools.

Comment (degorenko) [reply pkg/apis/ceph.rook.io/v1/types.go]: > If the set the name in a filesystem CR that already exists, and it is different than the existing generated pool name, it would certainly create a problem so we should add a check for that case. 

We can't add a check for that, since we are not tracking diff for spec (only showing it as simple diff string from comparer). Also, it is mostly same to situation, when user dropped old pool and added a new one named - real diff impossible to track. So, i'd rather to leave it up to user and warning them to avoid such situations, like add a different/new names for already present pools.

> If they add the name, maybe we should enforce that the name be set on all the pools in the list. This would allow them to be reordered, even for the existing default-named pools.

I dont think that it is needed. Since we already have an ordered list, user may just add new pools with predefined names and do not touch already present.

Again, my question is not answered: do we need to create cephfs data pools with cephfs name prefix for NamedPoolSpec, e.g. <cephfsName>-<nameFromSpec> ?

Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go]: > > If the set the name in a filesystem CR that already exists, and it is different than the existing generated pool name, it would certainly create a problem so we should add a check for that case.
> 
> We can't add a check for that, since we are not tracking diff for spec (only showing it as simple diff string from comparer). Also, it is mostly same to situation, when user dropped old pool and added a new one named - real diff impossible to track. So, i'd rather to leave it up to user and warning them to avoid such situations, like add a different/new names for already present pools.

Agreed, not sure how to check for that. 

> > If they add the name, maybe we should enforce that the name be set on all the pools in the list. This would allow them to be reordered, even for the existing default-named pools.
> 
> I dont think that it is needed. Since we already have an ordered list, user may just add new pools with predefined names and do not touch already present.

It's not really necessary, but it could help save some error scenarios or at least confusion around some pools having names and some not. We can skip the error check though, I'm fine with that. 

> Again, my question is not answered: do we need to create cephfs data pools with cephfs name prefix for NamedPoolSpec, e.g. - ?

This would certainly help ensure the pool names are unique when there are multiple filesystems with named pools, otherwise the pools could mistakenly overlap. My assumption with #9295 was that arbitrary pool names could be used, but sounds like you're interested in better names than data0, data1, ...

Btw I'm working on #9209 and the `NamedPoolSpec` will be useful for that implementation as well. I hope to have a PR open by tomorrow if not today. That change should set a foundation for what we want to do with this PR, so perhaps you wait for that to avoid too much implementation overlap.

Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go]: With #9363, hopefully this will be a much more straight forward implementation. The `NamedPoolSpec` is already passed down to the pool creation code, so it might basically look like:
- Change the types.go for dataPools to `NamedPoolSpec`
- In the filesystem.go, where you see `dataPoolNames[i]`, implement the name override if specified instead of using the generated name.

Comment (degorenko) [reply pkg/apis/ceph.rook.io/v1/types.go]: > With #9363, hopefully this will be a much more straight forward implementation. The NamedPoolSpec is already passed down to the pool creation code, so it might basically look like:
Change the types.go for dataPools to NamedPoolSpec

Ok, let's wait. Then i'll just rebase, once your PR is merged.

> In the filesystem.go, where you see dataPoolNames[i], implement the name override if specified instead of using the generated name.

Actually, it is already done. But again, should we continue to use real pool names as `<cephfsName>-<poolNameFromSpec>` or just `<poolNameFromSpec>`?

Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go]: > But again, should we continue to use real pool names as `<cephfsName>-<poolNameFromSpec>` or just `<poolNameFromSpec>`?

I'm thinking `<cephfsName>-<poolNameFromSpec>` to prevent conflicts between filesystems. wdyt?

Comment (degorenko) [reply pkg/apis/ceph.rook.io/v1/types.go]: Yeah, agree. I'll update.

**Comment by degorenko** [comment]
@travisn is there any reason to wait actually #9363 ? Defining NamedPoolSpec in current change is not complex. May be we can go ahead and merge since CI is passed and all comments resolved?

File: Documentation/ceph-filesystem-crd.md
Comment (travisn) [inline Documentation/ceph-filesystem-crd.md:34]: pool in the name seems redundant, how about removing the pool suffix from the examples?
```suggestion
    - name: replicated
```

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:644]:
```suggestion
	// Name of the pool
```

File: pkg/operator/ceph/file/filesystem.go
Comment (travisn) [inline pkg/operator/ceph/file/filesystem.go:282]: How about a unit test for this name generation method?

File: Documentation/ceph-filesystem-crd.md
Comment (travisn) [inline Documentation/ceph-filesystem-crd.md:194]:
```suggestion
* `dataPools`: The settings to create the filesystem data pools. Optionally, a pool name can be specified with the `name` field to override the generated name. If multiple pools are specified, Rook will add the pools to the filesystem. Assigning users or files to a pool is left as an exercise for the reader with the [CephFS documentation](http://docs.ceph.com/docs/master/cephfs/file-layouts/). The data pools can use replication or erasure coding. If erasure coding pools are specified, the cluster must be running with bluestore enabled on the OSDs.
```

**Review by travisn — CHANGES_REQUESTED** [review]
> @travisn is there any reason to wait actually #9363 ? Defining NamedPoolSpec in current change is not complex. May be we can go ahead and merge since CI is passed and all comments resolved?

Yes, we could go ahead with it, it's independent enough. Just a few more small comments...

File: Documentation/ceph-filesystem-crd.md
Comment (degorenko) [reply Documentation/ceph-filesystem-crd.md:34]: but since it is for example, i just wanted to mention *pool* suffix for storage class. Otherwise, in storage class we will have pool name just as myfs-replicated. Having myfs-replicated-pool (with pool name replicated-pool) a bit more user friendly, haven't it?

File: pkg/operator/ceph/file/filesystem.go
Comment (degorenko) [reply pkg/operator/ceph/file/filesystem.go:282]: Sure, let me add it

File: Documentation/ceph-filesystem-crd.md
Comment (degorenko) [reply Documentation/ceph-filesystem-crd.md:194]: > Optionally, a pool name can be specified with the `name` field to override the generated name.

How about:

Optionally, a pool name can be specified with the `name` field to override the default generated name.

File: Documentation/ceph-filesystem-crd.md
Comment (travisn) [reply Documentation/ceph-filesystem-crd.md:34]: In the storage class, the setting is: `pool: myfs-replicated-pool`, right? If the setting name is `pool`, isn't it clear it's a pool?

Comment (degorenko) [reply Documentation/ceph-filesystem-crd.md:34]: Okay, you win :)

File: deploy/examples/crds.yaml
Comment (BlaineEXE) [inline deploy/examples/crds.yaml:5800]: Should `dataPools` really be optional now?

File: deploy/examples/crds.yaml
Comment (BlaineEXE) [inline deploy/examples/crds.yaml:4971]: We have had upgrade problems with things being turned from `nullable` to non-nullable. Should we keep this?

File: deploy/examples/csi/cephfs/storageclass-ec.yaml
Comment (BlaineEXE) [inline deploy/examples/csi/cephfs/storageclass-ec.yaml:19]: There are also some examples in `nfs.yaml` and `nfs-test.yaml` IIRC. Those might have been gone for a bit, so rebasing and doing a search for `myfs-data0` again might help

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:645]: I'm still struggling a bit with calling this `name` since the name ends up being partly generated. At the end of the day, I'm not sure something like `generateName` is very useful here and that improving the docs is the best thing to do.

We still might want to make this struct CephFilesystem specific (i.e., not reusing the named pool spec for CephBlockPool) and having the doc text be something like "unique identifier for the pool within the filesystem. will be prefixed with the metadata name." We should also mention that we do most definitely recommend setting the pool name here since it helps resolve errors that could arise from altering the pool list.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:976]: I think this might be a carryover from the previous design?

File: Documentation/ceph-filesystem-crd.md
Comment (BlaineEXE) [inline Documentation/ceph-filesystem-crd.md:194]: I would like to be clearer here that while the `name` is optional and that we highly recommend setting the name to prevent issues that can arise from modifying the spec in a way that causes Rook to lose the original pool ordering. I would also suggest adding `name` as a field under `dataPools` here so that we can also document that the actual pool name stored in Ceph for each pool will be `<metadata.name>-<name>`. It will be easier to read the section just on `name` if it is separated. We can still mention here in the `dataPools` section that we recommend setting the `name` parameter on on pools that is documented below.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: deploy/examples/crds.yaml
Comment (travisn) [reply deploy/examples/crds.yaml:4971]: We really do require at least one data pool, so not sure there would be an impact here either way.

File: deploy/examples/crds.yaml
Comment (travisn) [reply deploy/examples/crds.yaml:5800]: keeping required makes sense

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:645]: I like the NamedPoolSpec to use both for the CephBlockPool and the CephFilesystem. In the Filesystem docs sounds good to just be clear that the filesystem name is prepended to the name.

File: deploy/examples/crds.yaml
Comment (degorenko) [reply deploy/examples/crds.yaml:5800]: Agreed, it still stays required

File: deploy/examples/csi/cephfs/storageclass-ec.yaml
Comment (degorenko) [reply deploy/examples/csi/cephfs/storageclass-ec.yaml:19]: `nsf.yaml` contains next:
```
  rados:
    # RADOS pool where NFS configs are stored.
    # In this example the data pool for the "myfs" filesystem is used.
    # If using the object store example, the data pool would be "my-store.rgw.buckets.data".
    # Note that this has nothing to do with where exported file systems or object stores live.
    pool: myfs-data0
```

So i guess, it is clear, why used such name, isn't it?

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (degorenko) [reply pkg/apis/ceph.rook.io/v1/types.go:645]: > In the Filesystem docs sounds good to just be clear that the filesystem name is prepended to the name.

Yeah, we may add note that name for pool will be prepended to the filesystem name.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (degorenko) [reply pkg/apis/ceph.rook.io/v1/types.go:976]: Looks like, i'll return nullable as it was before

**Review by travisn — APPROVED** [review]

File: Documentation/ceph-filesystem-crd.md
Comment (BlaineEXE) [inline Documentation/ceph-filesystem-crd.md:194]: I think users will be able to read the docs more easily if the `name` field has its own section. This way, we can be clear about other features of `dataPools` without creating a paragraph that is cumbersome to understand.

```suggestion
* `dataPools`: The settings to create the filesystem data pools. Optionally (and we highly recommend), a pool name can be specified with the `name` field to override the default generated name; see more below.If multiple pools are specified, Rook will add the pools to the filesystem. Assigning users or files to a pool is left as an exercise for the reader with the [CephFS documentation](http://docs.ceph.com/docs/master/cephfs/file-layouts/). The data pools can use replication or erasure coding. If erasure coding pools are specified, the cluster must be running with bluestore enabled on the OSDs.
  * `name`: (optional, and highly recommended) Override the default generated name of the pool. The final pool name will consist of the filesystem name and pool name, e.g., `<fsName>-<poolName>`. We highly recommend to specify `name` to prevent issues that can arise from modifying the spec in a way that causes Rook to lose the original pool ordering. 
```

File: deploy/examples/csi/cephfs/storageclass.yaml
Comment (BlaineEXE) [inline deploy/examples/csi/cephfs/storageclass.yaml:17]: We should change `myfs-data0` in other manifests (like `nfs.yaml` and `nfs-test.yaml`) also. Otherwise, the NFS examples won't successfully run with the new filesystem manifests.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
Just a couple small doc and example manifest updates. Thanks, @degorenko

File: Documentation/ceph-filesystem-crd.md
Comment (degorenko) [reply Documentation/ceph-filesystem-crd.md]: Done

File: deploy/examples/csi/cephfs/storageclass.yaml
Comment (degorenko) [reply deploy/examples/csi/cephfs/storageclass.yaml:17]: Ok, updated.

**Review by BlaineEXE — APPROVED** [review]

**Pull request merged**
